Pass assigner type to Pimpl

anosach-luxoft opened the pull request:
Current `Pimpl` class realization doesn't allow to choose copy&assign behavior. `Impl`s pointers swapping had been hardcoded as only available way. We have to extend `Pimpl` class with possibility to choose how it will copy&assign `Impl`s pointers. We have cases where we need `Pimpl`s with swap, copy and non-copy assignment policy. So this pull request provides such ability for `Pimpl` class by passing `Assigner` functor to `Pimpl` class as template parameter. Swap behavior will be kept as default to support present code

anosach-luxoft committed:
Pass assigner type to Pimpl

anosach-luxoft commented:
@pestOO @nk0leg @AGaliuzov please review

nk0leg commented on src/components/include/utils/pimpl.h:
```diff
@@ -43,26 +69,22 @@ namespace utils {
+ * @tparam Assigner Type of Assigner for Pimpl
  **/
-template <typename Impl>
+template <typename Impl, typename Assigner = SwapAssigner<Impl> >
```
Please check with MSVC 2010

anosach-luxoft replied:
@nk0leg  works fine with MSVC 2010

anosach-luxoft committed:
Make copy constructor and assign operator non-const

Assigner::operator() receives non-const pointers as
arguments because it's going to change them. As far
as we use Assigner::operator() in copy constructor
and assign operator we should keep their parameters
non-const too

anosach-luxoft commented:
@malirod see 29f4415

pestOO commented:
@anosach-luxoft, please provide PR inforamtion about changes: what have been implemented and what for.

pestOO commented:
@anosach-luxoft, please provide commit message information according agreements.

anosach-luxoft commented:
@malirod @pestOO please see 30b60a7. PR header has also been updated

pestOO commented:
Reviewed

anosach-luxoft committed:
Pass arguments by reference

anosach-luxoft commented:
@malirod PR has been rebased. Also f9c0546 has been added

anosach-luxoft committed:
Remove non-copy assigner as redundant
